providers/implementations/signature/sm2sig.c

```diff
@@ -0,0 +1,502 @@
+static OSSL_OP_signature_digest_sign_final_fn sm2sig_digest_sign_final;
+static OSSL_OP_signature_digest_verify_init_fn sm2sig_digest_signverify_init;
+static OSSL_OP_signature_digest_verify_update_fn sm2sig_digest_signverify_update;
+static OSSL_OP_signature_digest_verify_final_fn sm2sig_digest_verify_final;
+static OSSL_OP_signature_freectx_fn sm2sig_freectx;
+static OSSL_OP_signature_dupctx_fn sm2sig_dupctx;
+static OSSL_OP_signature_get_ctx_params_fn sm2sig_get_ctx_params;
+static OSSL_OP_signature_gettable_ctx_params_fn sm2sig_gettable_ctx_params;
+static OSSL_OP_signature_set_ctx_params_fn sm2sig_set_ctx_params;
+static OSSL_OP_signature_settable_ctx_params_fn sm2sig_settable_ctx_params;
+static OSSL_OP_signature_get_ctx_md_params_fn sm2sig_get_ctx_md_params;
+static OSSL_OP_signature_gettable_ctx_md_params_fn sm2sig_gettable_ctx_md_params;
+static OSSL_OP_signature_set_ctx_md_params_fn sm2sig_set_ctx_md_params;
+static OSSL_OP_signature_settable_ctx_md_params_fn sm2sig_settable_ctx_md_params;
+
+/*
+ * What's passed as an actual key is defined by the KEYMGMT interface.
+ * We happen to know that our KEYMGMT simply passes EC structures, so
+ * we use that here too.
+ */
+typedef struct {
+    OPENSSL_CTX *libctx;
+    EC_KEY *ec;
+
+    /*
+     * Flag to determine if the hash function can be changed (1) or not (0)
+     * Because it's dangerous to change during a DigestSign or DigestVerify
+     * operation, this flag is cleared by their Init function, and set again
+     * by their Final function.
+     */
+    unsigned int flag_allow_md : 1;
+
+    char mdname[OSSL_MAX_NAME_SIZE];
+
+    /* The Algorithm Identifier of the combined signature agorithm */
+    unsigned char aid[OSSL_MAX_ALGORITHM_ID_SIZE];
+    size_t  aid_len;
+
+    /* main digest */
+    EVP_MD *md;
+    EVP_MD_CTX *mdctx;
+    size_t mdsize;
+} PROV_SM2_CTX;
+
+/* TODO: it seems SM2 doesn't need this */
+static int sm2sig_get_md_nid(const EVP_MD *md)
+{
+    /*
+     * Because the EC library deals with NIDs, we need to translate.
+     * We do so using EVP_MD_is_a(), and therefore need a name to NID
+     * map.
+     */
+    static const OSSL_ITEM name_to_nid[] = {
+        { NID_sm2,   OSSL_DIGEST_NAME_SM3 },
+    };
+    size_t i;
+    int mdnid = NID_undef;
+
+    if (md == NULL)
+        goto end;
+
+    for (i = 0; i < OSSL_NELEM(name_to_nid); i++) {
+        if (EVP_MD_is_a(md, name_to_nid[i].ptr)) {
+            mdnid = (int)name_to_nid[i].id;
+            break;
+        }
+    }
+
+    if (mdnid == NID_undef)
+        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_DIGEST);
+
+ end:
+    return mdnid;
+}
+
+static void *sm2sig_newctx(void *provctx)
+{
+    PROV_SM2_CTX *psm2ctx = OPENSSL_zalloc(sizeof(PROV_SM2_CTX));
+
+    if (psm2ctx == NULL)
+        return NULL;
+
+    psm2ctx->libctx = PROV_LIBRARY_CONTEXT_OF(provctx);
+    /* don't allow to change MD, and in fact there is no such need */
+    psm2ctx->flag_allow_md = 0;
```

Comment:
So, can we just delete this field? It seems unnecessary.

Reply:
Right, if we decide to limit the algorithm to SM2 only.